pm/recipient_test.go

```diff
@@ -0,0 +1,851 @@
+
+	r, err := NewRecipient(b, v, ts, faceValue, winProb)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub broker with zero deposit and penalty escrow
+	b.SetDeposit(sender, big.NewInt(0))
+	b.SetPenaltyEscrow(sender, big.NewInt(0))
+
+	// Test zero deposit and penalty escrow error
+	ticket := newTicket(sender, params, 0)
+
+	won, err := r.ReceiveTicket(ticket, sig, params.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !won {
+		t.Fatal("expected valid winning ticket")
+	}
+
+	err = r.RedeemWinningTickets(ticket.RecipientRandHash.Hex())
+	if err == nil {
+		t.Error("expected zero deposit and penalty escrow error")
+	}
+	if err != nil && !strings.Contains(err.Error(), "zero deposit and penalty escrow") {
+		t.Errorf("expected zero deposit and penalty escrow error, got %v", err)
+	}
+}
+
+func TestRedeemWinningTickets_SingleTicket_RedeemError(t *testing.T) {
+	sender, b, v, ts, faceValue, winProb, sig := newRecipientFixture(t)
+	secret := [32]byte{3}
+
+	r := NewRecipientWithSecret(b, v, ts, secret, faceValue, winProb)
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub broker with zero deposit and penalty escrow and to fail redemption
+	b.SetDeposit(sender, big.NewInt(0))
+	b.SetPenaltyEscrow(sender, big.NewInt(500))
+	b.redeemShouldFail = true
+
+	// Test redeem error
+	ticket := newTicket(sender, params, 2)
+
+	won, err := r.ReceiveTicket(ticket, sig, params.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !won {
+		t.Fatal("expected valid winning ticket")
+	}
+
+	err = r.RedeemWinningTickets(ticket.RecipientRandHash.Hex())
+	if err == nil {
+		t.Error("expected ticket redemption error")
+	}
+
+	used, err := b.IsUsedTicket(ticket)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if used {
+		t.Error("expected non-used ticket")
+	}
+
+	recipientRand := genRecipientRand(sender, secret, params.Seed)
+
+	if _, ok := r.(*recipient).invalidRands.Load(recipientRand.String()); ok {
+		t.Error("expected not to invalidate recipientRand")
+	}
+
+	if _, ok := r.(*recipient).senderNonces[recipientRand.String()]; !ok {
+		t.Error("expected not to clear senderNonce memory")
+	}
+}
+
+func TestRedeemWinningTickets_SingleTicket(t *testing.T) {
+	sender, b, v, ts, faceValue, winProb, sig := newRecipientFixture(t)
+	secret := [32]byte{3}
+
+	r := NewRecipientWithSecret(b, v, ts, secret, faceValue, winProb)
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub broker with zero deposit and penalty escrow and to fail redemption
+	b.SetDeposit(sender, big.NewInt(0))
```

Comment:
Why should deposit be zero here?

Reply:
Updated in a72a3af to set up stub broker with non-zero deposit and penalty escrow for the test sender so that we only have to explicitly set deposit and penalty escrow to zero when testing for errors